src/stepFunctions/commands/visualizeStateMachine.ts

```diff
@@ -29,257 +29,339 @@ export interface messageObject {
+
+        return newVisualization.getPanel()
     }
 
-    const debouncedUpdate = debounce(sendUpdateMessage, 500)
-
-    // Set the initial html for the webpage
-    panel.webview.html = getWebviewContent(
-        ext.visualizationResourcePaths.webviewBodyScript.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.visualizationLibraryScript.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.visualizationLibraryCSS.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.stateMachineCustomThemeCSS.with({ scheme: 'vscode-resource' }),
-        {
-            inSync: localize(
-                'AWS.stepFunctions.graph.status.inSync',
-                'Previewing ASL document. <a href="" style="text-decoration:none;">View</a>'
-            ),
-            notInSync: localize('AWS.stepFunctions.graph.status.notInSync', 'Errors detected. Cannot preview.'),
-            syncing: localize('AWS.stepFunctions.graph.status.syncing', 'Rendering ASL graph...'),
+    private getExistingVisualization(documentToFind: vscode.TextDocument): AslVisualization | void {
+        for (const vis of this.managedVisualizations) {
+            if (vis.documentUri.path === documentToFind.uri.path) {
+                return vis
+            }
         }
-    )
 
-    // Add listener function to update the graph on document save
-    const updateOnSaveDisposable = vscode.workspace.onDidSaveTextDocument(async savedTextDocument => {
-        if (savedTextDocument && savedTextDocument.uri.path === documentUri.path) {
-            await sendUpdateMessage(savedTextDocument)
+        return
+    }
+}
+
+export class AslVisualization {
+    public readonly documentUri: vscode.Uri
+    public readonly webviewPanel: vscode.WebviewPanel
+    protected readonly disposables: vscode.Disposable[] = []
+    protected isPanelDisposed: boolean = false
+    private readonly onVisualizationDisposeEmitter = new vscode.EventEmitter<void>()
+
+    public constructor(textDocument: vscode.TextDocument) {
+        this.documentUri = textDocument.uri
+        this.webviewPanel = this.setupWebviewPanel(textDocument)
+    }
+
+    public getOnVisualizationDisposeEvent() {
+        return this.onVisualizationDisposeEmitter.event
+    }
+
+    public getPanel(): vscode.WebviewPanel | void {
+        if (this.webviewPanel && !this.isPanelDisposed) {
+            return this.webviewPanel
         }
-    })
+    }
 
-    const onDidCloseTextDocumentDisposable = vscode.workspace.onDidCloseTextDocument(documentWillSaveEvent => {
-        if (!trackedDocumentDoesExist(documentUri)) {
-            panel.dispose()
-            vscode.window.showInformationMessage(
-                localize(
-                    'AWS.stepfunctions.visualisation.errors.rename',
-                    'State machine visualization closed due to file renaming or closure.'
-                )
-            )
+    public getWebview(): vscode.Webview | void {
```

Comment:
getWebView and showPanel appear to duplicate logic that is performed in getPanel. It should be possible to have these methods directly call getPanel and operate from there. (returning "x or undefined" may help here, eg: this.getPanel?.reveal())